--- /Users/matejk/git/github/poco/dependencies/expat/src/xmlparse.c	2025-12-16 15:25:55
+++ /Users/matejk/git/github/poco/dependencies/expat/src/xmlparse.cpp	2025-12-16 15:30:43
@@ -104,7 +104,12 @@
 #include <stdint.h> /* uintptr_t */
 #include <math.h>   /* isnan */
 
-#ifdef _WIN32
+#define EXPAT_POCO 1
+
+#if defined(EXPAT_POCO)
+#  include "Poco/RandomStream.h"
+#  include "Poco/BinaryReader.h"
+#elif defined(_WIN32)
 #  define getpid GetCurrentProcessId
 #else
 #  include <sys/time.h>  /* gettimeofday() */
@@ -114,7 +119,7 @@
 #  include <errno.h>
 #endif
 
-#ifdef _WIN32
+#ifdef EXPAT_WIN32
 #  include "winconfig.h"
 #endif
 
@@ -146,7 +151,7 @@
 #if ! defined(HAVE_GETRANDOM) && ! defined(HAVE_SYSCALL_GETRANDOM)             \
     && ! defined(HAVE_ARC4RANDOM_BUF) && ! defined(HAVE_ARC4RANDOM)            \
     && ! defined(XML_DEV_URANDOM) && ! defined(_WIN32)                         \
-    && ! defined(XML_POOR_ENTROPY)
+    && ! defined(XML_POOR_ENTROPY) && ! defined(EXPAT_POCO)
 #  error You do not have support for any sources of high quality entropy \
     enabled.  For end user security, that is probably not what you want. \
     \
@@ -797,7 +802,7 @@
 
 #if XML_GE == 1
 static void
-expat_heap_stat(XML_Parser rootParser, char operator, XmlBigCount absDiff,
+expat_heap_stat(XML_Parser rootParser, char op, XmlBigCount absDiff,
                 XmlBigCount newTotal, XmlBigCount peakTotal, int sourceLine) {
   // NOTE: This can be +infinity or -nan
   const float amplification
@@ -806,7 +811,7 @@
       stderr,
       "expat: Allocations(%p): Direct " EXPAT_FMT_ULL("10") ", allocated %c" EXPAT_FMT_ULL(
           "10") " to " EXPAT_FMT_ULL("10") " (" EXPAT_FMT_ULL("10") " peak), amplification %8.2f (xmlparse.c:%d)\n",
-      (void *)rootParser, rootParser->m_accounting.countBytesDirect, operator,
+      (void *)rootParser, rootParser->m_accounting.countBytesDirect, op,
       absDiff, newTotal, peakTotal, (double)amplification, sourceLine);
 }
 
@@ -1034,6 +1039,9 @@
        ASCII_SLASH, ASCII_n,     ASCII_a,      ASCII_m,      ASCII_e,
        ASCII_s,     ASCII_p,     ASCII_a,      ASCII_c,      ASCII_e,
        '\0'};
+
+
+#if ! defined(EXPAT_POCO)
 
 /* To avoid warnings about unused functions: */
 #if ! defined(HAVE_ARC4RANDOM_BUF) && ! defined(HAVE_ARC4RANDOM)
@@ -1190,6 +1198,8 @@
 }
 
 #endif /* ! defined(HAVE_ARC4RANDOM_BUF) && ! defined(HAVE_ARC4RANDOM) */
+
+#endif /* ! defined(EXPAT_POCO) */
 
 static unsigned long
 ENTROPY_DEBUG(const char *label, unsigned long entropy) {
@@ -1204,7 +1214,12 @@
 generate_hash_secret_salt(XML_Parser parser) {
   unsigned long entropy;
   (void)parser;
-
+#if defined(EXPAT_POCO)
+  Poco::RandomInputStream rstr;
+  Poco::BinaryReader rrdr(rstr);
+  rrdr >> entropy;
+  return ENTROPY_DEBUG("RandomInputStream", entropy);
+#else
   /* "Failproof" high quality providers: */
 #if defined(HAVE_ARC4RANDOM_BUF)
   arc4random_buf(&entropy, sizeof(entropy));
@@ -1241,6 +1256,7 @@
                          entropy * (unsigned long)2305843009213693951ULL);
   }
 #endif
+#endif /* defined(EXPAT_POCO) */
 }
 
 static unsigned long
@@ -1368,7 +1384,7 @@
       parser = (XML_Parser)((char *)sizeAndParser + sizeof(size_t)
                             + EXPAT_MALLOC_PADDING);
 #else
-    parser = memsuite->malloc_fcn(sizeof(struct XML_ParserStruct));
+    parser = (XML_Parser)memsuite->malloc_fcn(sizeof(struct XML_ParserStruct));
     if (parser != NULL) {
 #endif
       mtemp = (XML_Memory_Handling_Suite *)&(parser->m_mem);
@@ -1444,20 +1460,20 @@
   parser->m_bufferLim = NULL;
 
   parser->m_attsSize = INIT_ATTS_SIZE;
-  parser->m_atts = MALLOC(parser, parser->m_attsSize * sizeof(ATTRIBUTE));
+  parser->m_atts = (ATTRIBUTE*)MALLOC(parser, parser->m_attsSize * sizeof(ATTRIBUTE));
   if (parser->m_atts == NULL) {
     FREE(parser, parser);
     return NULL;
   }
 #ifdef XML_ATTR_INFO
-  parser->m_attInfo = MALLOC(parser, parser->m_attsSize * sizeof(XML_AttrInfo));
+  parser->m_attInfo = (XML_AttrInfo*)MALLOC(parser, parser->m_attsSize * sizeof(XML_AttrInfo));
   if (parser->m_attInfo == NULL) {
     FREE(parser, parser->m_atts);
     FREE(parser, parser);
     return NULL;
   }
 #endif
-  parser->m_dataBuf = MALLOC(parser, INIT_DATA_BUF_SIZE * sizeof(XML_Char));
+  parser->m_dataBuf = (XML_Char*)MALLOC(parser, INIT_DATA_BUF_SIZE * sizeof(XML_Char));
   if (parser->m_dataBuf == NULL) {
     FREE(parser, parser->m_atts);
 #ifdef XML_ATTR_INFO
@@ -2598,7 +2614,7 @@
       }
       // NOTE: We are avoiding MALLOC(..) here to leave limiting
       //       the input size to the application using Expat.
-      newBuf = parser->m_mem.malloc_fcn(bufferSize);
+      newBuf = (char*)parser->m_mem.malloc_fcn(bufferSize);
       if (newBuf == 0) {
         parser->m_errorCode = XML_ERROR_NO_MEMORY;
         return NULL;
@@ -3143,7 +3159,7 @@
       return XML_FALSE;
     bufSize = nameLen + rawNameLen;
     if (bufSize > (size_t)(tag->bufEnd - tag->buf)) {
-      char *temp = REALLOC(parser, tag->buf, bufSize);
+      char *temp = (char*)REALLOC(parser, tag->buf, bufSize);
       if (temp == NULL)
         return XML_FALSE;
       /* if tag->name.str points to tag->buf (only when namespace
@@ -3469,10 +3485,10 @@
         tag = parser->m_freeTagList;
         parser->m_freeTagList = parser->m_freeTagList->parent;
       } else {
-        tag = MALLOC(parser, sizeof(TAG));
+        tag = (TAG*)MALLOC(parser, sizeof(TAG));
         if (! tag)
           return XML_ERROR_NO_MEMORY;
-        tag->buf = MALLOC(parser, INIT_TAG_BUF_SIZE);
+        tag->buf = (char*)MALLOC(parser, INIT_TAG_BUF_SIZE);
         if (! tag->buf) {
           FREE(parser, tag);
           return XML_ERROR_NO_MEMORY;
@@ -3505,7 +3521,7 @@
           }
           bufSize = (int)(tag->bufEnd - tag->buf) << 1;
           {
-            char *temp = REALLOC(parser, tag->buf, bufSize);
+            char *temp = (char*)REALLOC(parser, tag->buf, bufSize);
             if (temp == NULL)
               return XML_ERROR_NO_MEMORY;
             tag->buf = temp;
@@ -3884,7 +3900,7 @@
     }
 #endif
 
-    temp = REALLOC(parser, parser->m_atts,
+    temp = (ATTRIBUTE*)REALLOC(parser, parser->m_atts,
                    parser->m_attsSize * sizeof(ATTRIBUTE));
     if (temp == NULL) {
       parser->m_attsSize = oldAttsSize;
@@ -3903,7 +3919,7 @@
     }
 #  endif
 
-    temp2 = REALLOC(parser, parser->m_attInfo,
+    temp2 = (XML_AttrInfo*)REALLOC(parser, parser->m_attInfo,
                     parser->m_attsSize * sizeof(XML_AttrInfo));
     if (temp2 == NULL) {
       parser->m_attsSize = oldAttsSize;
@@ -4080,7 +4096,7 @@
       }
 #endif
 
-      temp = REALLOC(parser, parser->m_nsAtts, nsAttsSize * sizeof(NS_ATT));
+      temp = (NS_ATT*)REALLOC(parser, parser->m_nsAtts, nsAttsSize * sizeof(NS_ATT));
       if (! temp) {
         /* Restore actual size of memory in m_nsAtts */
         parser->m_nsAttsPower = oldNsAttsPower;
@@ -4261,7 +4277,7 @@
     }
 #endif
 
-    uri = MALLOC(parser, (n + EXPAND_SPARE) * sizeof(XML_Char));
+    uri = (XML_Char*)MALLOC(parser, (n + EXPAND_SPARE) * sizeof(XML_Char));
     if (! uri)
       return XML_ERROR_NO_MEMORY;
     binding->uriAlloc = n + EXPAND_SPARE;
@@ -4508,7 +4524,7 @@
 #endif
 
       XML_Char *temp
-          = REALLOC(parser, b->uri, sizeof(XML_Char) * (len + EXPAND_SPARE));
+          = (XML_Char*)REALLOC(parser, b->uri, sizeof(XML_Char) * (len + EXPAND_SPARE));
       if (temp == NULL)
         return XML_ERROR_NO_MEMORY;
       b->uri = temp;
@@ -4516,7 +4532,7 @@
     }
     parser->m_freeBindingList = b->nextTagBinding;
   } else {
-    b = MALLOC(parser, sizeof(BINDING));
+    b = (BINDING*)MALLOC(parser, sizeof(BINDING));
     if (! b)
       return XML_ERROR_NO_MEMORY;
 
@@ -4534,7 +4550,7 @@
     }
 #endif
 
-    b->uri = MALLOC(parser, sizeof(XML_Char) * (len + EXPAND_SPARE));
+    b->uri = (XML_Char*)MALLOC(parser, sizeof(XML_Char) * (len + EXPAND_SPARE));
     if (! b->uri) {
       FREE(parser, b);
       return XML_ERROR_NO_MEMORY;
@@ -4954,7 +4970,7 @@
     if (parser->m_unknownEncodingHandler(parser->m_unknownEncodingHandlerData,
                                          encodingName, &info)) {
       ENCODING *enc;
-      parser->m_unknownEncodingMem = MALLOC(parser, XmlSizeOfUnknownEncoding());
+      parser->m_unknownEncodingMem = (void*)MALLOC(parser, XmlSizeOfUnknownEncoding());
       if (! parser->m_unknownEncodingMem) {
         if (info.release)
           info.release(info.data);
@@ -5905,7 +5921,7 @@
               return XML_ERROR_NO_MEMORY;
             }
 
-            char *const new_connector = REALLOC(
+            char *const new_connector = (char*)REALLOC(
                 parser, parser->m_groupConnector, parser->m_groupSize *= 2);
             if (new_connector == NULL) {
               parser->m_groupSize /= 2;
@@ -5925,14 +5941,14 @@
             }
 #endif
 
-            int *const new_scaff_index = REALLOC(
+            int *const new_scaff_index = (int*)REALLOC(
                 parser, dtd->scaffIndex, parser->m_groupSize * sizeof(int));
             if (new_scaff_index == NULL)
               return XML_ERROR_NO_MEMORY;
             dtd->scaffIndex = new_scaff_index;
           }
         } else {
-          parser->m_groupConnector = MALLOC(parser, parser->m_groupSize = 32);
+          parser->m_groupConnector = (char*)MALLOC(parser, parser->m_groupSize = 32);
           if (! parser->m_groupConnector) {
             parser->m_groupSize = 0;
             return XML_ERROR_NO_MEMORY;
@@ -6093,7 +6109,7 @@
           //       applications that are not using XML_FreeContentModel but
           //       plain free(..) or .free_fcn() to free the content model's
           //       memory are safe.
-          XML_Content *content = parser->m_mem.malloc_fcn(sizeof(XML_Content));
+          XML_Content *content = (XML_Content*)parser->m_mem.malloc_fcn(sizeof(XML_Content));
           if (! content)
             return XML_ERROR_NO_MEMORY;
           content->quant = XML_CQUANT_NONE;
@@ -6370,7 +6386,7 @@
     openEntity = *freeEntityList;
     *freeEntityList = openEntity->next;
   } else {
-    openEntity = MALLOC(parser, sizeof(OPEN_INTERNAL_ENTITY));
+    openEntity = (OPEN_INTERNAL_ENTITY*)MALLOC(parser, sizeof(OPEN_INTERNAL_ENTITY));
     if (! openEntity)
       return XML_ERROR_NO_MEMORY;
   }
@@ -7170,7 +7186,7 @@
     if (type->allocDefaultAtts == 0) {
       type->allocDefaultAtts = 8;
       type->defaultAtts
-          = MALLOC(parser, type->allocDefaultAtts * sizeof(DEFAULT_ATTRIBUTE));
+          = (DEFAULT_ATTRIBUTE*)MALLOC(parser, type->allocDefaultAtts * sizeof(DEFAULT_ATTRIBUTE));
       if (! type->defaultAtts) {
         type->allocDefaultAtts = 0;
         return 0;
@@ -7195,7 +7211,7 @@
       }
 #endif
 
-      temp = REALLOC(parser, type->defaultAtts,
+      temp = (DEFAULT_ATTRIBUTE*)REALLOC(parser, type->defaultAtts,
                      (count * sizeof(DEFAULT_ATTRIBUTE)));
       if (temp == NULL)
         return 0;
@@ -7488,7 +7504,7 @@
 
 static DTD *
 dtdCreate(XML_Parser parser) {
-  DTD *p = MALLOC(parser, sizeof(DTD));
+  DTD *p = (DTD*)MALLOC(parser, sizeof(DTD));
   if (p == NULL)
     return p;
   poolInit(&(p->pool), parser);
@@ -7672,7 +7688,7 @@
       }
 #endif
       newE->defaultAtts
-          = MALLOC(parser, oldE->nDefaultAtts * sizeof(DEFAULT_ATTRIBUTE));
+          = (DEFAULT_ATTRIBUTE*)MALLOC(parser, oldE->nDefaultAtts * sizeof(DEFAULT_ATTRIBUTE));
       if (! newE->defaultAtts) {
         return 0;
       }
@@ -7834,7 +7850,7 @@
     /* table->size is a power of 2 */
     table->size = (size_t)1 << INIT_POWER;
     tsize = table->size * sizeof(NAMED *);
-    table->v = MALLOC(table->parser, tsize);
+    table->v = (NAMED**)MALLOC(table->parser, tsize);
     if (! table->v) {
       table->size = 0;
       return NULL;
@@ -7874,7 +7890,7 @@
       }
 
       size_t tsize = newSize * sizeof(NAMED *);
-      NAMED **newV = MALLOC(table->parser, tsize);
+      NAMED **newV = (NAMED**)MALLOC(table->parser, tsize);
       if (! newV)
         return NULL;
       memset(newV, 0, tsize);
@@ -7903,7 +7919,7 @@
       }
     }
   }
-  table->v[i] = MALLOC(table->parser, createSize);
+  table->v[i] = (NAMED*)MALLOC(table->parser, createSize);
   if (! table->v[i])
     return NULL;
   memset(table->v[i], 0, createSize);
@@ -8150,7 +8166,7 @@
     if (bytesToAllocate == 0)
       return XML_FALSE;
 
-    temp = REALLOC(pool->parser, pool->blocks, bytesToAllocate);
+    temp = (BLOCK*)REALLOC(pool->parser, pool->blocks, bytesToAllocate);
     if (temp == NULL)
       return XML_FALSE;
     pool->blocks = temp;
@@ -8190,7 +8206,7 @@
     if (bytesToAllocate == 0)
       return XML_FALSE;
 
-    tem = MALLOC(pool->parser, bytesToAllocate);
+    tem = (BLOCK*)MALLOC(pool->parser, bytesToAllocate);
     if (! tem)
       return XML_FALSE;
     tem->size = blockSize;
@@ -8221,7 +8237,7 @@
       return -1;
     }
 #endif
-    dtd->scaffIndex = MALLOC(parser, parser->m_groupSize * sizeof(int));
+    dtd->scaffIndex = (int*)MALLOC(parser, parser->m_groupSize * sizeof(int));
     if (! dtd->scaffIndex)
       return -1;
     dtd->scaffIndex[0] = 0;
@@ -8249,13 +8265,13 @@
       }
 #endif
 
-      temp = REALLOC(parser, dtd->scaffold,
+      temp = (CONTENT_SCAFFOLD*)REALLOC(parser, dtd->scaffold,
                      dtd->scaffSize * 2 * sizeof(CONTENT_SCAFFOLD));
       if (temp == NULL)
         return -1;
       dtd->scaffSize *= 2;
     } else {
-      temp = MALLOC(parser, INIT_SCAFFOLD_ELEMENTS * sizeof(CONTENT_SCAFFOLD));
+      temp = (CONTENT_SCAFFOLD*)MALLOC(parser, INIT_SCAFFOLD_ELEMENTS * sizeof(CONTENT_SCAFFOLD));
       if (temp == NULL)
         return -1;
       dtd->scaffSize = INIT_SCAFFOLD_ELEMENTS;
@@ -8312,7 +8328,7 @@
   // NOTE: We are avoiding MALLOC(..) here to so that
   //       applications that are not using XML_FreeContentModel but plain
   //       free(..) or .free_fcn() to free the content model's memory are safe.
-  ret = parser->m_mem.malloc_fcn(allocsize);
+  ret = (XML_Content*)parser->m_mem.malloc_fcn(allocsize);
   if (! ret)
     return NULL;
 
@@ -8445,7 +8461,7 @@
   charsRequired++;
 
   /* Now allocate space for the copy */
-  result = MALLOC(parser, charsRequired * sizeof(XML_Char));
+  result = (XML_Char*)MALLOC(parser, charsRequired * sizeof(XML_Char));
   if (result == NULL)
     return NULL;
   /* Copy the original into place */
